proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,90 @@
+This will allow one to easily identify all the packages released as part of a single product version release as well as request bulk-processing for a set of packages which belong to the same release.
+For example, all packages published as part of .NET 8 Preview 5 release will be staged together and then also can be deprecated together by a single call to NuGet APIs.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `published`, which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of hte time developers would simply pick the latest version of all these packages, which somtimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+There is a new field in the package metadata named `release`, which indicates the unique release of the product the package belongs to.
+Server implementations are expected to treat each `release` as scoped to an account or similar concept, depending on the server's implementation details.
+For example, on nuget.org, User1 and User2 can both use a `release` named `V1.0.0`, but when User1 operates on their `V1.0.0` release, it does not affect User2's release with the same name.
```

Comment:
This concept needs to be reconciled with multiple owners on a package. I take it that the set of owners on a package containing this `release` all share the same `release` entity? Theoretically only the entity that pushed the specific package version (i.e. the organization of the pusher or the specific co-owner that pushed the package) could relate to the `release` entity. This is a big ambiguous of a point I'm making here, so let's take an example.

If user `joelverhagen` pushes release `v1.0.0` to his `Knapcode.TorSharp` package but `joelverhagen` co-owns a package `Contoso.MyLib` with `artak`, does Artak share the `v1.0.0` release entity? Probably not. But what if `v1.0.0` was used on `Contoso.MyLib`, the shared package. Then `joelverhagen` now has two potentially unrelated `v1.0.0` release entities. 

Mutli-owner packages introduce these complexities.

Reply:
This is indeed the most challenging part of the proposal. Including in the updated one that I'm working on now. Still trying to come up with a good plan here.